src/main/java/com/iexec/worker/chain/ContributionAuthorizationService.java

```diff
@@ -0,0 +1,67 @@
+package com.iexec.worker.chain;
+
+import com.iexec.common.chain.*;
+import com.iexec.common.contract.generated.IexecHubABILegacy;
+import com.iexec.common.replicate.ReplicateStatusCause;
+import com.iexec.common.security.Signature;
+import com.iexec.common.utils.BytesUtils;
+import com.iexec.common.utils.HashUtils;
+import com.iexec.common.utils.SignatureUtils;
+import com.iexec.worker.config.PublicConfigurationService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import javax.annotation.PostConstruct;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Optional;
+
+import static com.iexec.common.replicate.ReplicateStatusCause.*;
+
+
+@Slf4j
+@Service
+public class ContributionAuthorizationService {
+
+    private PublicConfigurationService publicConfigurationService;
+    private HashMap<String, ContributionAuthorization> contributionAuthorizations;
+    private String corePublicAddress;
+
+    public ContributionAuthorizationService(PublicConfigurationService publicConfigurationService) {
+        this.publicConfigurationService = publicConfigurationService;
+    }
+
+    @PostConstruct
+    public void initIt() {
+        corePublicAddress = publicConfigurationService.getSchedulerPublicAddress();
+        contributionAuthorizations = new HashMap<>();
```

Comment:
Maybe we should use a ConcurrentHashMap to avoid any multi-threading issue.

Reply:
Wanted but not implemented, thanks

I changed too the `tasksUsingCpu HashSet` to an `ConcurentHashSet`.